akka-typed/src/main/scala/akka/typed/javadsl/BehaviorChain.scala

```diff
@@ -0,0 +1,243 @@
+
+  /**
+   * Add a new case to the signal handling without compile time type check.
+   *
+   * Should normally not be used, but when matching on class with generic type
+   * argument it can be useful, e.g. <code>GenMsg.class</code> and <code>(ActorContext<Message> ctx, GenMsg&lt;String&gt; list) -> {...}</code>
+   *
+   * @param type type of signal to match
+   * @param handler action to apply when the type matches
+   * @return a new behavior with the specified handling appended
+   */
+  def signalUnchecked[M <: Signal](`type`: Class[_ <: Signal], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorChain[T] =
+    new BehaviorChain[T](messageHandlers, signalHandlers :+ Case[T, Signal](`type`, None, (ctx: ActorContext[T], signal: Signal) ⇒ handler.apply(ctx, signal.asInstanceOf[M])))
+
+  /**
+   * Add a new case to the signal handling matching equal signals.
+   *
+   * @param signal the signal to compare to
+   * @param handler action to apply when the message matches
+   * @return a new behavior with the specified handling appended
+   */
+  def signalEquals(signal: Signal, handler: Function[ActorContext[T], Behavior[T]]): BehaviorChain[T] =
+    new BehaviorChain[T](messageHandlers, signalHandlers :+ Case[T, Signal](signal.getClass, Some(_.equals(signal)), (ctx: ActorContext[T], _: Signal) ⇒ handler.apply(ctx)))
+
+}
+
+object BehaviorChain {
+  def create[T]: BehaviorChain[T] = new BehaviorChain[T](Nil, Nil)
+
+  import scala.language.existentials
+
+  /** INTERNAL API */
+  @InternalApi
+  private[javadsl] final case class Case[BT, MT](`type`: Class[_ <: MT], predicate: Option[MT ⇒ Boolean], handler: (ActorContext[BT], MT) ⇒ Behavior[BT])
+
+  /**
+   * Start a new behavior chain starting with this case.
+   *
+   * @param type type of message to match
+   * @param handler action to apply if the type matches
+   * @tparam T type of behavior to create
+   * @tparam M type of message to match
+   * @return a new behavior with the specified handling appended
+   */
+  def message[T, M <: T](`type`: Class[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorChain[T] =
+    BehaviorChain.create[T].message(`type`, handler)
+
+  /**
+   * Start a new behavior chain starting with this predicated case.
+   *
+   * @param type type of message to match
+   * @param test a predicate that will be evaluated on the argument if the type matches
+   * @param handler action to apply if the type matches and the predicate returns true
+   * @tparam T type of behavior to create
+   * @tparam M type of message to match
+   * @return a new behavior with the specified handling appended
+   */
+  def message[T, M <: T](`type`: Class[M], test: Predicate[M], handler: Function2[ActorContext[T], M, Behavior[T]]): BehaviorChain[T] =
```

Comment:
I'm not sure these "static" methods are worth it, let's see when we have decided how the `BehaviorChain` (builder) can be created